actors/builtin/miner/deadline_state.go

```diff
@@ -444,3 +446,161 @@ func (dl *Deadline) popExpiredPartitions(store adt.Store, until abi.ChainEpoch,
 
 	return popped, modified, nil
 }
+
+// RemovePartitions removes the specified partitions, shifting the remaining
+// ones to the left, and returning the live and dead sectors they contained.
+//
+// Returns an error if any of the partitions contained faulty sectors or early
+// terminations.
+func (dl *Deadline) RemovePartitions(store adt.Store, partitions *bitfield.BitField, quant QuantSpec) (
+	live, dead *abi.BitField, removedPower PowerPair, err error,
+) {
+	oldPartitions, err := dl.PartitionsArray(store)
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to load partitions: %w", err)
+	}
+
+	partitionCount := oldPartitions.Length()
+	toCompact, err := partitions.AllMap(partitionCount)
+	if err != nil {
+		// TODO: This may be an illegal argument error if we have too many partitions.
+		// https://github.com/filecoin-project/specs-actors/issues/597
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to expand partitions into map: %w", err)
+	}
+
+	// Nothing to do.
+	if len(toCompact) == 0 {
+		return bitfield.NewFromSet(nil), bitfield.NewFromSet(nil), NewPowerPairZero(), nil
+	}
+
+	for partIdx := range toCompact { //nolint:nomaprange
```

Comment:
I assume this is fine, right? Is it a problem if different miners fail the same transaction with different error messages?